rrdtool/rrd_mts_similarity.c

```diff
@@ -0,0 +1,483 @@
+    mts_data[i] = malloc(sizeof(rrd_value_t));
+
+  for(i=0; i<n_file; i++)
+  {
+    if(rrdms->rfs[i].path != NULL)
+    { 
+      if(rrd_fetch_r(rrdms->rfs[i].path, "AVERAGE", &start, &end, &step, &ds_cnt, &names, &data) != 0) {
+        printf("Unable to extract data from rrd %s\n", rrdms->rfs[i].path);
+        return;
+      }
+      
+      mts_data[j] = data;
+      j++;
+    }else{
+      not_found++;
+    }
+  }
+  
+  num_points = (end-start)/step;
+
+  if((s = ndpi_alloc_data_analysis(num_points)) == NULL)
+    return;
+
+  ndpi_init_bin(&rrdms->mts_b, ndpi_bin_family32, num_points);
+  
+  double value,acc;
+
+  /* Step 1 - Compute average and stddev */
+  for(t=start+1, j=0; t<end; t+=step, j++) {
+  value = 0;
+  acc = 0;
+    /* For each Time Series we took the t-th point */
+    for(i=0; i<n_file-not_found; i++){
+     
+      value = (double)*mts_data[i]++;		
+      if(!isnan(value)) acc += value;  			
+    }
+    
+    /* Multivariate Time Series takes the average of the values found */
+    ndpi_data_add_value(s, acc/(n_file-not_found));
+    ndpi_set_bin(&rrdms->mts_b, j, acc/(n_file-not_found));
+  }
+
+  /* Compute MTS's average and stddev */
+  rrdms->mts_average = ndpi_data_average(s);
+  rrdms->mts_stddev  = ndpi_data_stddev(s);
+
+  /* Step 2 - Bin analysis */
+  ndpi_free_data_analysis(s, 1);
+    
+  free(mts_data);
+}
+
+/* *************************************************** */
+
+int circles_touch(int x1, int r1, int x2, int r2) {
+  int radius_sum = r1+r2;
+  int x_diff     = abs(x1 - x2);
+
+  return((radius_sum < x_diff) ? 0 : 1);
+}
+
+/* *************************************************** */
+
+void find_rrd_similarities(rrd_multifile_stats *rrdms, int *num_rrds, u_int num_tot_rrds, int n_file, int num_host, char **filename) {
+  u_int i, j, k, num_similar_rrds, num_potentially_zero_equal;
+
+  for(k=0; k<n_file; k++) {
+    num_similar_rrds = 0;
+    num_potentially_zero_equal = 0;
+  
+    for(i=0; i<num_host; i++) {
+      for(j=i+1; j<num_host; j++) {
+      /*
+	Average is the circle center, and stddev is the radius
+	if circles touch each other then there is a chance that
+	the two rrds are similar
+      */
+      if((rrdms[i].rfs[k].path != NULL) && (rrdms[j].rfs[k].path != NULL))
+      {
+        if((rrdms[i].rfs[k].average == 0) && (rrdms[i].rfs[k].average == rrdms[j].rfs[k].average)) {
+	  if(!skip_zero)
+	      printf("%s [%.1f/%.1f]  - %s [%.1f/%.1f] are alike\n",
+		     rrdms[i].rfs[k].path, rrdms[i].rfs[k].average, rrdms[i].rfs[k].stddev,
+		     rrdms[j].rfs[k].path, rrdms[j].rfs[k].average, rrdms[j].rfs[k].stddev);
+	  
+	  num_potentially_zero_equal++;
+        } else if(circles_touch(rrdms[i].rfs[k].average, rrdms[i].rfs[k].stddev, rrdms[j].rfs[k].average, rrdms[j].rfs[k].stddev)
+	  	) {
+	  float similarity = ndpi_bin_similarity(&rrdms[i].rfs[k].b, &rrdms[j].rfs[k].b, 0, similarity_threshold);
+
+	  if((similarity >= 0) && (similarity < similarity_threshold)) {
```

Comment:
The cumulative similarity is not considered, but only individual. Not what I expect

Reply:
Fixed in commit with ID bcdca2f